File: libs/aifw-js/libner.js
import { env, AutoModelForTokenClassification } from '@xenova/transformers';
import * as Transformers from '@xenova/transformers';

let MODELS_BASE = '';

export function initEnv({ wasmBase = '/wasm/', modelsBase = '', threads, simd, fetchFn } = {}) {
  env.allowLocalModels = true;
  env.useBrowserCache = true;
  env.backends.onnx.wasm.wasmPaths = wasmBase;
  MODELS_BASE = modelsBase || '';
  if (MODELS_BASE) {
    const modelsBase = MODELS_BASE.endsWith('/') ? MODELS_BASE.slice(0, -1) : MODELS_BASE;
    env.localModelPath = modelsBase; // transformers will fetch from `${env.localModelPath}/${modelId}/...`
  }
  if (typeof fetchFn === 'function') {
    env.fetch = fetchFn;
  }
  if (typeof threads === 'number' && threads > 0) env.backends.onnx.wasm.numThreads = threads;
  if (typeof simd === 'boolean') env.backends.onnx.wasm.simd = simd;
}

export const SUPPORTED = new Set([
  'Xenova/distilbert-base-cased-finetuned-conll03-english',
  'gagan3012/bert-tiny-finetuned-ner',
  'dslim/distilbert-NER',
  'funstory-ai/neurobert-mini',
  'boltuix/NeuroBERT-Mini',
  "hfl/minirbt-h256",
  'dmis-lab/TinyPubMedBERT-v1.0',
  'boltuix/NeuroBERT-Small',
  'ckiplab/bert-tiny-chinese-ner',
]);

// Models that explicitly require/benefit from BertTokenizerFast
const PREFER_BERT_FAST = new Set([
  'ckiplab/bert-tiny-chinese-ner',
]);

function localDirFor(modelId) {
  // Rely on env.localModelPath as base; return only the model id
  return `${modelId}`;
}

export class TokenClassificationPipeline {
  constructor({ model, tokenizer = null, processor = null }) {
    this.model = model;
    this.tokenizer = tokenizer;
    this.processor = processor;
  }

  async run(text, { ignore_labels = ['O'], offsetText, tokenTransform } = {}) {
    const enc = this.tokenizer([text], { padding: true, truncation: true, add_special_tokens: true });
    const idsRow = enc.input_ids[0];
    const seqLen = idsRow.dims[0];

    let tokensPlain = [];
    const seqIndexToPlainIndex = new Array(seqLen).fill(-1);
    for (let j = 0; j < seqLen; j++) {
      const id = idsRow[j].item();
      const word = this.tokenizer.decode([id], { skip_special_tokens: true });
      if (word) {
        const plainIdx = tokensPlain.length;
        seqIndexToPlainIndex[j] = plainIdx;
        tokensPlain.push(word.startsWith('##') ? word.slice(2) : word);
      }
    }

    if (typeof tokenTransform === 'function') {
      const transformed = new Array(tokensPlain.length);
      for (let i = 0; i < tokensPlain.length; i++) transformed[i] = tokenTransform(tokensPlain[i]) || tokensPlain[i];
      tokensPlain = transformed;
    }

    const baseTextForOffsets = typeof offsetText === 'string' ? offsetText : text;
    const offsets = computeOffsetsFromTokens(baseTextForOffsets, tokensPlain);
    const outputs = await this.model(enc);
    const logits = outputs.logits;
    const dims = logits.dims || [];
    const numLabels = dims[2] || 0;
    const data = logits.data;
    const id2label = this.model.config.id2label;

    const items = [];
    for (let j = 0; j < seqLen; j++) {
      if (!numLabels || !data) break;
      const base = j * numLabels;
      if (base + numLabels > data.length) break;
      let maxIdx = 0, maxVal = -Infinity;
      for (let k = 0; k < numLabels; k++) {
        const v = data[base + k];
        if (v > maxVal) { maxVal = v; maxIdx = k; }
      }
      let sumExp = 0;
      for (let k = 0; k < numLabels; k++) sumExp += Math.exp(data[base + k] - maxVal);
      const score = sumExp > 0 ? Math.exp(data[base + maxIdx] - maxVal) / sumExp : 0;
      const entity = id2label ? id2label[maxIdx] : `LABEL_${maxIdx}`;
      if (ignore_labels.includes(entity)) continue;

      const plainIndex = seqIndexToPlainIndex[j];
      if (plainIndex < 0) continue;

      const word = tokensPlain[plainIndex] || '';
      const off = offsets[plainIndex] || [null, null];
      items.push({ entity, score, index: j, word, start: off[0], end: off[1] });
    }

    return mergeSubwordItems(items);
  }
}

export async function buildNerPipeline(modelId, { quantized = true, preferWebGPU = true } = {}) {
  if (!SUPPORTED.has(modelId)) throw new Error(`Unsupported model: ${modelId}`);
  const base = localDirFor(modelId);
  const modelPath = `${base}`;
  // Prefer BertTokenizerFast when requested and available, fallback to AutoTokenizer
  const TokenizerClass = PREFER_BERT_FAST.has(modelId)
    ? (Transformers.BertTokenizerFast || Transformers.AutoTokenizer)
    : Transformers.AutoTokenizer;
  const tokenizer = await TokenizerClass.from_pretrained(modelPath);
  const device = [];
  if (preferWebGPU && typeof navigator !== 'undefined' && 'gpu' in navigator) device.push('webgpu');
  device.push('wasm');
  const model = await AutoModelForTokenClassification.from_pretrained(modelPath, { quantized: !!quantized, device });
  return new TokenClassificationPipeline({ model, tokenizer });
}

export function computeOffsetsFromTokens(text, tokens) {
  const offsets = new Array(tokens.length);
  const lowerText = text.toLowerCase();
  const strippedInfo = buildStrippedMap(lowerText);
  const strippedText = strippedInfo.stripped;
  const mapStrippedToOrig = strippedInfo.map;

  let cursorLower = 0;
  let cursorStripped = 0;

  for (let i = 0; i < tokens.length; i++) {
    const full = String(tokens[i] ?? '');
    const raw = full.startsWith('##') ? full.slice(2) : full;
    if (!raw) { offsets[i] = [cursorLower, cursorLower]; continue; }

    const tokLower = raw.toLowerCase();
    let p = lowerText.indexOf(tokLower, cursorLower);
    if (p === -1) p = lowerText.indexOf(tokLower);
    if (p !== -1) {
      const s = p;
      const e = p + raw.length;
      offsets[i] = [s, e];
      cursorLower = e;
      cursorStripped = findStrippedIndexAtOrAfter(mapStrippedToOrig, e);
      continue;
    }

    const tokStripped = stripAccents(tokLower);
    if (tokStripped) {
      let sp = strippedText.indexOf(tokStripped, cursorStripped);
      if (sp === -1) sp = strippedText.indexOf(tokStripped);
      if (sp !== -1) {
        const startOrig = mapStrippedToOrig[sp] ?? 0;
        const after = sp + tokStripped.length;
        const endOrig = after < mapStrippedToOrig.length ? mapStrippedToOrig[after] : text.length;
        offsets[i] = [startOrig, endOrig];
        cursorLower = endOrig;
        cursorStripped = after;
        continue;
      }
    }

    offsets[i] = [cursorLower, cursorLower];
  }
  return offsets;
}

export function mergeSubwordItems(items) {
  if (!Array.isArray(items) || items.length === 0) return items;
  const out = [];
  let cur = null;
  let count = 0;
  const core = (e) => (e.startsWith('B-') || e.startsWith('I-') || e.startsWith('E-')) ? e.slice(2) : e;

  for (const it of items) {
    if (!cur) { cur = { ...it }; count = 1; continue; }
    const sameEntity = core(cur.entity) === core(it.entity);
    const contiguous = cur.end === it.start;
    if (sameEntity && contiguous) {
      cur.word = cur.word + it.word;
      cur.end = it.end;
      cur.score = (cur.score * count + it.score) / (count + 1);
      count += 1;
    } else {
      out.push(cur);
      cur = { ...it };
      count = 1;
    }
  }
  if (cur) out.push(cur);
  return out;
}

function stripAccents(s) {
  try { return s.normalize('NFD').replace(/\p{M}+/gu, ''); } catch { return s; }
}

function buildStrippedMap(s) {
  const outChars = [];
  const map = [];
  let i = 0;
  while (i < s.length) {
    const cp = s.codePointAt(i);
    const char = String.fromCodePoint(cp);
    const norm = char.normalize('NFD');
    for (let j = 0; j < norm.length; j++) {
      const ch = norm[j];
      if (/\p{M}/u.test(ch)) continue;
      if (isConnectorPunct(ch)) continue;
      outChars.push(ch);
      map.push(i);
    }
    i += char.length;
  }
  return { stripped: outChars.join(''), map };
}

function isConnectorPunct(ch) {
  return /[-'`\u2010-\u2015\u2212\u00B7\u30FB\u2043\u2219]/u.test(ch);
}

function findStrippedIndexAtOrAfter(map, origIndex) {
  let lo = 0, hi = map.length;
  while (lo < hi) {
    const mid = (lo + hi) >>> 1;
    if (map[mid] < origIndex) lo = mid + 1; else hi = mid;
  }
  return lo;
}

// Build a WASM buffer of NerRecogEntity (wasm32 C layout):
// struct {
//   uint8 entity_type;    // recog_entity.zig::EntityType
//   uint8 entity_tag;     // recog_entity.zig::EntityBioTag
//   uint16 _pad;          // padding to align to 4 bytes
//   float score;
//   uint32 index;
//   uint32 start;
//   uint32 end;
// } // total 20 bytes
export function buildNerEntitiesBuffer(wasm, items, jsText) {
  if (!items?.length) return { ptr: 0, count: 0, owned: [], byteSize: 0 };
  const structSize = 20;
  const count = items.length >>> 0;
  const total = structSize * count;
  const arrPtr = wasm.aifw_malloc(total);
  if (!arrPtr) throw new Error('aifw_malloc ner entities failed');
  const dv = new DataView(wasm.memory.buffer);

  const isText = typeof jsText === 'string';
  const textLen = isText ? jsText.length : Number(jsText || 0);
  const utf8Map = isText ? computeUtf8OffsetMap(jsText) : null;

  // Keep in sync with core/recog_entity.zig
  const EntityType = {
    None: 0,
    PHYSICAL_ADDRESS: 1,
    EMAIL_ADDRESS: 2,
    ORGANIZATION: 3,
    USER_MAME: 4,
    PHONE_NUMBER: 5,
    BANK_NUMBER: 6,
    PAYMENT: 7,
    VERIFICATION_CODE: 8,
    PASSWORD: 9,
    RANDOM_SEED: 10,
    PRIVATE_KEY: 11,
    URL_ADDRESS: 12,
  };
  const BioTag = { None: 0, Begin: 1, Inside: 2 };

  // Return tuple: [coreLabel, bioTag]
  const toCoreAndTag = (e) => {
    const s = String(e || '');
    if (s.startsWith('B-')) return [s.slice(2), BioTag.Begin];
    if (s.startsWith('S-')) return [s.slice(2), BioTag.Begin]; // single-token entity -> treat as Begin
    if (s.startsWith('I-')) return [s.slice(2), BioTag.Inside];
    if (s.startsWith('E-')) return [s.slice(2), BioTag.Inside]; // end-token -> treat as Inside
    if (s) return [s, BioTag.None];
    return ['MISC', BioTag.None];
  };
  const toEntityType = (core) => {
    switch (core) {
      case 'PER': case 'PERSON': return EntityType.USER_MAME;
      case 'ORG': return EntityType.ORGANIZATION;
      case 'LOC': case 'GPE': case 'FAC': case 'ADDRESS':
        return EntityType.PHYSICAL_ADDRESS;
      case 'MISC': return EntityType.None;
      default: return EntityType.None;
    }
  };

  for (let i = 0; i < count; i++) {
    const it = items[i];
    let s = Math.max(0, Math.min(textLen, Number(it.start || 0)));
    let e = Math.max(s, Math.min(textLen, Number(it.end || 0)));
    // Convert character indices to UTF-8 byte indices for the core
    if (utf8Map) {
      const sByte = utf8Map[s] ?? 0;
      const eByte = utf8Map[e] ?? utf8Map[textLen] ?? sByte;
      s = sByte;
      e = Math.max(sByte, eByte);
    }
    const base = arrPtr + i * structSize;
    const [entityCore, tagVal] = toCoreAndTag(it.entity);
    const entityTypeVal = toEntityType(entityCore);

    dv.setUint8(base + 0, entityTypeVal);
    dv.setUint8(base + 1, tagVal);
    // base+2..+3 padding (leave as 0)
    dv.setFloat32(base + 4, Number(it.score || 0), true);
    dv.setUint32(base + 8, Number(it.index || 0) >>> 0, true);
    dv.setUint32(base + 12, s >>> 0, true);
    dv.setUint32(base + 16, e >>> 0, true);
  }
  return { ptr: arrPtr, count, owned: [], byteSize: total };
}

// Build a map from JS string index (UTF-16 code unit index) to UTF-8 byte offset
function computeUtf8OffsetMap(text) {
  const map = new Array(text.length + 1);
  let bytePos = 0;
  let i = 0;
  while (i < text.length) {
    map[i] = bytePos;
    const cp = text.codePointAt(i);
    const cuLen = cp > 0xFFFF ? 2 : 1;
    let utf8Len = 0;
    if (cp <= 0x7F) utf8Len = 1; else if (cp <= 0x7FF) utf8Len = 2; else if (cp <= 0xFFFF) utf8Len = 3; else utf8Len = 4;
    bytePos += utf8Len;
    i += cuLen;
  }
  map[text.length] = bytePos;
  return map;
}

File: libs/aifw-py/libner.py
"""
NER utilities for aifw-py.

This module provides a minimal interface analogous to aifw-js/libner.js:
- init_env(opts)
- build_ner_pipeline(model_id, options)

Internally, it uses Python transformers (tokenizer + model) to generate token classification logits,
and reproduces the offset mapping/merging behavior similar to aifw-js/libner.js.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple
import logging

from transformers import AutoTokenizer
import numpy as np
import os
from typing import Any

logger = logging.getLogger(__name__)


_ENV_STATE: Dict[str, Any] = {
    "modelsBase": None,
}

_MODELS_BASE: Optional[str] = None


def init_env(opts: Optional[Dict[str, Any]] = None) -> None:
    """
    Initialize underlying environment.
    Kept for API parity with aifw-js; options are accepted but not required.
    """
    global _ENV_STATE, _MODELS_BASE
    opts = opts or {}
    _ENV_STATE = {
        "modelsBase": opts.get("modelsBase"),
    }
    mb = _ENV_STATE.get("modelsBase")
    _MODELS_BASE = str(mb) if isinstance(mb, str) and mb else None
    logger.info("[aifw-py] init_env completed; modelsBase=%s", _MODELS_BASE or "(unset)")


@dataclass
class NerItem:
    entity: str
    score: float
    index: int
    word: str
    start: int
    end: int


class TokenClassificationPipelinePy:
    """
    Lightweight analogue to the JS TokenClassificationPipeline.
    It emits items with fields: entity, score, index, word, start, end.
    """

    def __init__(self, ort_session: Optional[Any], tokenizer: Optional[AutoTokenizer], id2label: Dict[int, str], lang_hint: Optional[str] = None):
        self.ort_session = ort_session
        self.tokenizer = tokenizer
        self.id2label = id2label
        self._lang_hint = (lang_hint or "en").lower()

    async def run(self, text: str, opts: Optional[Dict[str, Any]] = None) -> List[NerItem]:
        """
        Re-implement token classification similar to JS:
        - Tokenize text
        - Run ONNX model (quantized) via onnxruntime
        - Softmax per-token to get label + score
        - Compute offsets on baseTextForOffsets using tokens (with optional tokenTransform)
        - Merge contiguous items of same entity
        """
        opts = opts or {}
        ignore_labels: List[str] = opts.get("ignore_labels", ["O"])
        offset_text: Optional[str] = opts.get("offsetText")
        token_transform = opts.get("tokenTransform")

        if self.ort_session is None or self.tokenizer is None:
            return []  # disabled pipeline

        enc = self.tokenizer(
            [text],
            return_offsets_mapping=True,
            return_tensors=None,
            padding=True,
            truncation=True,
            add_special_tokens=True,
        )
        input_ids = np.array(enc["input_ids"], dtype=np.int64)  # [1, seq_len]
        attention_mask = np.array(enc.get("attention_mask", [[1] * len(enc["input_ids"][0])]), dtype=np.int64)
        token_type_ids = enc.get("token_type_ids")
        if token_type_ids is not None:
            token_type_ids = np.array(token_type_ids, dtype=np.int64)

        # Prepare feed dict (robust to differing input names)
        feed = {}
        for inp in self.ort_session.get_inputs():
            name = inp.name
            if name in ("input_ids", "input_ids:0"):
                feed[name] = input_ids
            elif name in ("attention_mask", "attention_mask:0"):
                feed[name] = attention_mask
            elif name in ("token_type_ids", "token_type_ids:0") and token_type_ids is not None:
                feed[name] = token_type_ids
            elif name not in feed:
                # Fallback: try mapping by known keys
                if "input_ids" in name:
                    feed[name] = input_ids
                elif "attention_mask" in name:
                    feed[name] = attention_mask
                elif "token_type_ids" in name and token_type_ids is not None:
                    feed[name] = token_type_ids

        outputs = self.ort_session.run(None, feed)
        # Assume first output is logits: [1, seq_len, num_labels]
        logits = outputs[0]
        if logits.ndim == 3:
            logits = logits[0]
        seq_len = logits.shape[0]
        num_labels = logits.shape[1]

        # Plain tokens from ids (skip special, map '##' prefixes for BERT)
        tokens_plain: List[str] = []
        seq_index_to_plain: List[int] = [-1] * seq_len
        # Build tokens via ids; rely on tokenizer's convert_ids_to_tokens
        ids_row = input_ids[0]
        for j in range(seq_len):
            token_str = self.tokenizer.convert_ids_to_tokens(int(ids_row[j]))
            # Skip special tokens that often begin with [ or are empty
            if not token_str or token_str.startswith("[") and token_str.endswith("]"):
                continue
            if token_str.startswith("##"):
                token_str = token_str[2:]
            plain_idx = len(tokens_plain)
            seq_index_to_plain[j] = plain_idx
            tokens_plain.append(token_str)

        if callable(token_transform):
            tokens_plain = [token_transform(t) or t for t in tokens_plain]

        base_text_for_offsets = offset_text if isinstance(offset_text, str) else text
        offsets = compute_offsets_from_tokens(base_text_for_offsets, tokens_plain)

        items_raw: List[Tuple[int, str, float, str, int, int]] = []
        for j in range(seq_len):
            # softmax
            row = logits[j]
            max_idx = int(np.argmax(row))
            # numerical stable softmax
            max_val = float(np.max(row))
            exp = np.exp(row - max_val)
            sum_exp = float(np.sum(exp))
            score = float(exp[max_idx] / sum_exp) if sum_exp > 0 else 0.0
            entity = self.id2label.get(max_idx, f"LABEL_{max_idx}")
            if entity in ignore_labels:
                continue
            plain_idx = seq_index_to_plain[j]
            if plain_idx < 0 or plain_idx >= len(tokens_plain):
                continue
            word = tokens_plain[plain_idx]
            off = offsets[plain_idx] if 0 <= plain_idx < len(offsets) else (0, 0)
            items_raw.append((j, entity, score, word, int(off[0]), int(off[1])))

        # Merge contiguous same-entity segments
        items: List[NerItem] = []
        cur: Optional[NerItem] = None
        count = 0
        def core_label(s: str) -> str:
            if s.startswith("B-") or s.startswith("I-") or s.startswith("E-") or s.startswith("S-"):
                return s[2:]
            return s
        for (idx_j, ent, sc, wd, st, ed) in items_raw:
            if cur is None:
                cur = NerItem(entity=ent, score=sc, index=idx_j, word=wd, start=st, end=ed)
                count = 1
                continue
            same = core_label(cur.entity) == core_label(ent)
            contiguous = (cur.end == st)
            if same and contiguous:
                cur.word = cur.word + wd
                cur.end = ed
                cur.score = (cur.score * count + sc) / (count + 1)
                count += 1
            else:
                items.append(cur)
                cur = NerItem(entity=ent, score=sc, index=idx_j, word=wd, start=st, end=ed)
                count = 1
        if cur is not None:
            items.append(cur)
        return items


# Supported model ids for parity.
SUPPORTED_MODELS = {
    "funstory-ai/neurobert-mini",
    "ckiplab/bert-tiny-chinese-ner",
    "Xenova/distilbert-base-cased-finetuned-conll03-english",
    "gagan3012/bert-tiny-finetuned-ner",
    "dslim/distilbert-NER",
    "boltuix/NeuroBERT-Mini",
    "hfl/minirbt-h256",
    "dmis-lab/TinyPubMedBERT-v1.0",
    "boltuix/NeuroBERT-Small",
}


async def build_ner_pipeline(model_id: str, options: Optional[Dict[str, Any]] = None) -> TokenClassificationPipelinePy:
    """
    Build a transformers-based token classification pipeline, aligned with JS expectations.
    """
    base = _MODELS_BASE
    lang_hint = "zh" if ("chinese" in model_id.lower() or "ckiplab" in model_id.lower()) else "en"

    # If modelsBase not provided, or local path missing, return a no-op pipeline
    class _NoopPipe(TokenClassificationPipelinePy):
        def __init__(self):
            super().__init__(model=None, tokenizer=None, lang_hint=lang_hint)
        async def run(self, text: str, opts: Optional[Dict[str, Any]] = None) -> List[NerItem]:
            return []

    if not base:
        logger.info("[aifw-py] modelsBase not set; NER pipeline will be disabled (regex-only).")
        return _NoopPipe()
    model_dir = os.path.join(base.rstrip("/"), model_id)
    if not os.path.isdir(model_dir):
        logger.warning("[aifw-py] model path not found: %s; NER pipeline disabled (regex-only).", model_dir)
        return _NoopPipe()
    onnx_path = os.path.join(model_dir, "onnx", "model_quantized.onnx")
    if not os.path.isfile(onnx_path):
        logger.warning("[aifw-py] ONNX model not found: %s; NER pipeline disabled (regex-only).", onnx_path)
        return _NoopPipe()
    try:
        tokenizer = AutoTokenizer.from_pretrained(model_dir, local_files_only=True, trust_remote_code=False)
    except Exception as e:
        logger.warning("[aifw-py] failed to load tokenizer at %s: %s; NER pipeline disabled.", model_dir, e)
        return _NoopPipe()

    # Load id2label from config.json to recover labels like "B-PER", "I-ORG", ...
    id2label: Dict[int, str] = {}
    config_path = os.path.join(model_dir, "config.json")
    if os.path.isfile(config_path):
        import json
        try:
            with open(config_path, "r", encoding="utf-8") as f:
                cfg_json = json.load(f)
            raw = cfg_json.get("id2label") or {}
            if isinstance(raw, dict):
                # Keys may be int or stringified int
                id2label = {int(k): str(v) for k, v in raw.items()}
        except Exception as e:
            logger.warning("[aifw-py] failed to load id2label from %s: %s", config_path, e)
    if not id2label:
        logger.warning("[aifw-py] id2label not found in %s; NER entities will use LABEL_x and may not map to PII types.", config_path)

    # Defer onnxruntime import to runtime; make optional
    try:
        import onnxruntime as ort  # type: ignore
        # Create ONNX Runtime session (CPU)
        sess_options = ort.SessionOptions()
        ort_session = ort.InferenceSession(onnx_path, sess_options, providers=["CPUExecutionProvider"])
        return TokenClassificationPipelinePy(ort_session=ort_session, tokenizer=tokenizer, id2label=id2label, lang_hint=lang_hint)
    except Exception as e:
        logger.warning("[aifw-py] failed to create onnxruntime session for %s: %s; NER pipeline disabled.", onnx_path, e)
        return _NoopPipe()


# ---- Helpers ported from JS offset logic ----
def strip_accents(s: str) -> str:
    try:
        import unicodedata
        return "".join(ch for ch in unicodedata.normalize("NFD", s) if not unicodedata.combining(ch))
    except Exception:
        return s


def is_connector_punct(ch: str) -> bool:
    import re
    return re.match(r"[-'`\u2010-\u2015\u2212\u00B7\u30FB\u2043\u2219]", ch) is not None


def build_stripped_map(s: str) -> Tuple[str, List[int]]:
    out_chars: List[str] = []
    mapping: List[int] = []
    i = 0
    while i < len(s):
        ch = s[i]
        norm = ch
        try:
            import unicodedata
            norm = unicodedata.normalize("NFD", ch)
        except Exception:
            pass
        for c in norm:
            try:
                import unicodedata
                if unicodedata.combining(c):
                    continue
            except Exception:
                pass
            if is_connector_punct(c):
                continue
            out_chars.append(c)
            mapping.append(i)
        i += 1
    return "".join(out_chars), mapping


def find_stripped_index_at_or_after(mapping: List[int], orig_index: int) -> int:
    lo, hi = 0, len(mapping)
    while lo < hi:
        mid = (lo + hi) >> 1
        if mapping[mid] < orig_index:
            lo = mid + 1
        else:
            hi = mid
    return lo


def compute_offsets_from_tokens(text: str, tokens: List[str]) -> List[Tuple[int, int]]:
    offsets: List[Tuple[int, int]] = [(0, 0)] * len(tokens)
    lower_text = text.lower()
    stripped, map_stripped_to_orig = build_stripped_map(lower_text)
    cursor_lower = 0
    cursor_stripped = 0
    for i, full in enumerate(tokens):
        raw = (full or "")
        if raw.startswith("##"):
            raw = raw[2:]
        if not raw:
            offsets[i] = (cursor_lower, cursor_lower)
            continue
        tok_lower = raw.lower()
        p = lower_text.find(tok_lower, cursor_lower)
        if p == -1:
            p = lower_text.find(tok_lower)
        if p != -1:
            s = p
            e = p + len(raw)
            offsets[i] = (s, e)
            cursor_lower = e
            cursor_stripped = find_stripped_index_at_or_after(map_stripped_to_orig, e)
            continue
        tok_stripped = strip_accents(tok_lower)
        if tok_stripped:
            sp = stripped.find(tok_stripped, cursor_stripped)
            if sp == -1:
                sp = stripped.find(tok_stripped)
            if sp != -1:
                start_orig = map_stripped_to_orig[sp] if sp < len(map_stripped_to_orig) else 0
                after = sp + len(tok_stripped)
                end_orig = map_stripped_to_orig[after] if after < len(map_stripped_to_orig) else len(text)
                offsets[i] = (start_orig, end_orig)
                cursor_lower = end_orig
                cursor_stripped = after
                continue
        offsets[i] = (cursor_lower, cursor_lower)
    return offsets


